File: MaWPhotos/src/main/java/us/mikeandwan/photos/ui/mode/ModeSelectionActivity.java
package us.mikeandwan.photos.ui.mode;

import android.app.NotificationManager;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import com.google.android.material.snackbar.Snackbar;
import androidx.appcompat.widget.Toolbar;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.ExpandableListView;
import android.widget.ImageView;
import android.widget.SimpleExpandableListAdapter;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.Flowable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;
import us.mikeandwan.photos.MawApplication;
import us.mikeandwan.photos.R;
import us.mikeandwan.photos.di.ActivityComponent;
import us.mikeandwan.photos.di.DaggerActivityComponent;
import us.mikeandwan.photos.models.ApiCollection;
import us.mikeandwan.photos.models.Category;
import us.mikeandwan.photos.services.AuthStateManager;
import us.mikeandwan.photos.services.DataServices;
import us.mikeandwan.photos.services.PhotoListType;
import us.mikeandwan.photos.services.UpdateCategoriesJobScheduler;
import us.mikeandwan.photos.ui.BaseActivity;
import us.mikeandwan.photos.ui.HasComponent;
import us.mikeandwan.photos.ui.categories.CategoryListActivity;
import us.mikeandwan.photos.ui.login.LoginActivity;
import us.mikeandwan.photos.ui.photos.PhotoListActivity;
import us.mikeandwan.photos.ui.receiver.PhotoReceiverActivity;
import us.mikeandwan.photos.ui.settings.SettingsActivity;


public class ModeSelectionActivity extends BaseActivity implements HasComponent<ActivityComponent> {
    private static final String KEY_NAME = "NAME";
    private static final String KEY_TYPE = "TYPE";
    private static final long FOUR_HOURS_IN_MILLIS = 4 * 60 * 60 * 1000;  // 4 hours

    private final CompositeDisposable _disposables = new CompositeDisposable();
    private final List<Map<String, String>> _groupData = new ArrayList<>();
    private final List<List<Map<String, String>>> _childData = new ArrayList<>();
    private final List<Map<String, String>> _yearChildren = new ArrayList<>();
    private SimpleExpandableListAdapter _adapter;
    private List<Integer> _yearList;
    private MenuItem _refreshMenuItem;
    private ActivityComponent _activityComponent;
    private MawApplication _app;

    @BindView(R.id.toolbar) Toolbar _toolbar;
    @BindView(R.id.modeExpandableListView) ExpandableListView _modeExpandableListView;

    @Inject DataServices _dataServices;
    @Inject AuthStateManager _authStateManager;
    @Inject UpdateCategoriesJobScheduler _updateScheduler;
    @Inject SharedPreferences _sharedPrefs;


    public ActivityComponent getComponent() {
        return _activityComponent;
    }


    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_mode_selection);
        ButterKnife.bind(this);

        _app = (MawApplication) getApplication();

        _activityComponent = DaggerActivityComponent.builder()
                .applicationComponent(getApplicationComponent())
                .activityModule(getActivityModule())
                .build();

        _activityComponent.inject(this);

        updateToolbar(_toolbar, null);

        _modeExpandableListView.setOnChildClickListener(this::onItemClicked);

        _adapter = new SimpleExpandableListAdapter(this,
                _groupData,
                android.R.layout.simple_expandable_list_item_1,
                new String[]{KEY_NAME},
                new int[]{android.R.id.text1},
                _childData,
                android.R.layout.simple_expandable_list_item_1,
                new String[]{KEY_NAME},
                new int[]{android.R.id.text1});

        _updateScheduler.schedule(false, FOUR_HOURS_IN_MILLIS);

        initModeList();
        resetNotifications();
    }


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.mode_selection, menu);

        _refreshMenuItem = menu.findItem(R.id.action_force_sync);

        return true;
    }


    @Override
    public void onRestart() {
        // we do this on restart, so that if a new year was pulled during a refresh of the category
        // screen, or poller, this page will show the latest years we have available automatically
        prepareYearChildren();

        _adapter.notifyDataSetChanged();

        super.onRestart();
    }


    @Override
    protected void onDestroy() {
        super.onDestroy();
        _disposables.clear(); // do not send event after activity has been destroyed
    }


    public void onMenuItemSettings(MenuItem menuItem) {
        Intent intent = new Intent(this, SettingsActivity.class);
        startActivity(intent);
    }


    public void onMenuItemForceSync(MenuItem menuItem) {
        forceSync();
    }


    public void onViewUploadQueueClick(MenuItem menuItem) {
        Intent intent = new Intent(this, PhotoReceiverActivity.class);
        startActivity(intent);
    }


    public void onWipeCache(MenuItem menuItem) {
        _disposables.add(
            Flowable.fromCallable(() -> {
                    _dataServices.wipeCache();
                    return true;
                })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        this::onWipeComplete,
                        ex -> {
                            stopSyncAnimation();
                            handleApiException(ex);
                        })
        );
    }


    public void onReauthenticate(MenuItem menuItem) {
        _authStateManager.replace(null);

        Intent intent = new Intent(this, LoginActivity.class);
        startActivity(intent);

        finish();
    }


    private void resetNotifications() {
        NotificationManager mgr = (NotificationManager)getSystemService(NOTIFICATION_SERVICE);
        mgr.cancel(0);
        _app.setNotificationCount(0);
    }


    private void onWipeComplete(boolean b) {
        Snackbar.make(_toolbar, "Wipe complete.", Snackbar.LENGTH_SHORT).show();
    }


    private boolean onItemClicked(ExpandableListView parent, View v, int groupPosition, int childPosition, long id) {
        Intent intent;

        Map<String, String> child = _childData.get(groupPosition).get(childPosition);
        PhotoListType type = PhotoListType.valueOf(child.get(KEY_TYPE));

        if (type == PhotoListType.ByCategory) {
            intent = new Intent(parent.getContext(), CategoryListActivity.class);
            intent.putExtra("YEAR", Integer.parseInt(child.get(KEY_NAME)));
        } else {
            intent = new Intent(parent.getContext(), PhotoListActivity.class);
            intent.putExtra("NAME", child.get(KEY_NAME));
        }

        intent.putExtra("TYPE", type.toString());

        startActivity(intent);

        return true;
    }


    private void forceSync() {
        startSyncAnimation();

        _disposables.add(
            Flowable.fromCallable(() -> _dataServices.getRecentCategories())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                    this::onSyncComplete,
                    ex -> {
                        stopSyncAnimation();
                        handleApiException(ex);
                })
        );
    }


    private void onSyncComplete(ApiCollection<Category> categories) {
        stopSyncAnimation();

        long count = categories.getCount();

        if(count == 0) {
            Snackbar.make(_modeExpandableListView, "No updates available.", Snackbar.LENGTH_SHORT).show();
            return;
        }

        // TODO: simplify if/when we update api
        /*
        categories
                .stream()
                .filter(x -> !_yearList.contains(x.getYear()))
                .collect(Collectors.toList());
        */

        List<Integer> newYears = new ArrayList<>();

        for(Category cat : categories.getItems()) {
            if(!_yearList.contains(cat.getYear())) {
                newYears.add(cat.getYear());
            }
        }

        String year = "year";
        String cat = "category";

        if(count != 1) {
            cat = "categories";
        }

        if(newYears.size() != 1) {
            year = "years";
        }

        Snackbar.make(_modeExpandableListView, count + " new " + cat + " found. " + newYears.size() + " new " + year + " added.", Snackbar.LENGTH_SHORT).show();

        if(newYears.size() > 0) {
            prepareYearChildren();

            _adapter.notifyDataSetChanged();
        }
    }


    private void prepareYearChildren() {
        List<Integer> years = _dataServices.getPhotoYears();

        if (_yearList == null || years.size() != _yearList.size()) {
            _yearList = years;

            _yearChildren.clear();

            for (Integer year : _yearList) {
                _yearChildren.add(createChild(year.toString(), PhotoListType.ByCategory));
            }
        }
    }


    private void initModeList() {
        List<Map<String, String>> children;

        // by year
        prepareYearChildren();
        _groupData.add(createGroup("Year"));
        _childData.add(_yearChildren);

        // random
        children = new ArrayList<>();
        _groupData.add(createGroup("Random"));
        _childData.add(children);

        children.add(createChild("Surprise Me!", PhotoListType.Random));

        _modeExpandableListView.setAdapter(_adapter);
    }


    private Map<String, String> createGroup(String name) {
        Map<String, String> group = new HashMap<>();
        group.put(KEY_NAME, name);

        return group;
    }


    private Map<String, String> createChild(String name, PhotoListType type) {
        Map<String, String> child = new HashMap<>();
        child.put(KEY_NAME, name);
        child.put(KEY_TYPE, type.toString());

        return child;
    }


    private void startSyncAnimation() {
        ImageView iv = (ImageView) getLayoutInflater().inflate(R.layout.refresh_indicator, _toolbar, false);

        Animation rotation = AnimationUtils.loadAnimation(this, R.anim.refresh_rotate);
        rotation.setRepeatCount(Animation.INFINITE);
        iv.startAnimation(rotation);

        _refreshMenuItem.setActionView(iv);
    }


    private void stopSyncAnimation() {
        View v = _refreshMenuItem.getActionView();

        if(v != null) {
            v.clearAnimation();
            _refreshMenuItem.setActionView(null);
        }
    }
}

File: MaWPhotos/src/main/java/us/mikeandwan/photos/ui/mode/ModeSelectionActivity.kt
package us.mikeandwan.photos.ui.mode

import android.app.NotificationManager
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import android.widget.ExpandableListView
import android.widget.ImageView
import android.widget.SimpleExpandableListAdapter
import com.google.android.material.snackbar.Snackbar
import dagger.hilt.android.AndroidEntryPoint
import io.reactivex.Flowable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import us.mikeandwan.photos.MawApplication
import us.mikeandwan.photos.R
import us.mikeandwan.photos.databinding.ActivityModeSelectionBinding
import us.mikeandwan.photos.models.ApiCollection
import us.mikeandwan.photos.models.Category
import us.mikeandwan.photos.services.AuthStateManager
import us.mikeandwan.photos.services.DataServices
import us.mikeandwan.photos.services.PhotoListType
import us.mikeandwan.photos.services.UpdateCategoriesJobScheduler
import us.mikeandwan.photos.ui.BaseActivity
import us.mikeandwan.photos.ui.categories.CategoryListActivity
import us.mikeandwan.photos.ui.login.LoginActivity
import us.mikeandwan.photos.ui.photos.PhotoListActivity
import us.mikeandwan.photos.ui.receiver.PhotoReceiverActivity
import us.mikeandwan.photos.ui.settings.SettingsActivity
import java.util.*
import javax.inject.Inject

@AndroidEntryPoint
class ModeSelectionActivity : BaseActivity() {
    private val _disposables = CompositeDisposable()
    private val _groupData: MutableList<Map<String, String?>> = ArrayList()
    private val _childData: MutableList<List<Map<String, String?>>> = ArrayList()
    private val _yearChildren: MutableList<Map<String, String?>> = ArrayList()
    private var _adapter: SimpleExpandableListAdapter? = null
    private var _yearList: List<Int>? = null
    private var _refreshMenuItem: MenuItem? = null
    private var _app: MawApplication? = null

    private lateinit var binding: ActivityModeSelectionBinding

    @Inject lateinit var _dataServices: DataServices
    @Inject lateinit  var _authStateManager: AuthStateManager
    @Inject lateinit var _updateScheduler: UpdateCategoriesJobScheduler
    @Inject lateinit var _sharedPrefs: SharedPreferences

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityModeSelectionBinding.inflate(layoutInflater)
        setContentView(binding.root)

        _app = application as MawApplication
        updateToolbar(binding.toolbar, null)
        binding.modeExpandableListView.setOnChildClickListener { parent: ExpandableListView, v: View, groupPosition: Int, childPosition: Int, id: Long ->
            onItemClicked(
                parent,
                v,
                groupPosition,
                childPosition,
                id
            )
        }
        _adapter = SimpleExpandableListAdapter(
            this,
            _groupData,
            android.R.layout.simple_expandable_list_item_1,
            arrayOf(KEY_NAME),
            intArrayOf(android.R.id.text1),
            _childData,
            android.R.layout.simple_expandable_list_item_1,
            arrayOf(KEY_NAME),
            intArrayOf(android.R.id.text1)
        )
        _updateScheduler.schedule(false, FOUR_HOURS_IN_MILLIS)
        initModeList()
        resetNotifications()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.mode_selection, menu)
        _refreshMenuItem = menu.findItem(R.id.action_force_sync)
        return true
    }

    public override fun onRestart() {
        // we do this on restart, so that if a new year was pulled during a refresh of the category
        // screen, or poller, this page will show the latest years we have available automatically
        prepareYearChildren()
        _adapter!!.notifyDataSetChanged()
        super.onRestart()
    }

    override fun onDestroy() {
        super.onDestroy()
        _disposables.clear() // do not send event after activity has been destroyed
    }

    fun onMenuItemSettings(menuItem: MenuItem?) {
        val intent = Intent(this, SettingsActivity::class.java)
        startActivity(intent)
    }

    fun onMenuItemForceSync(menuItem: MenuItem?) {
        forceSync()
    }

    fun onViewUploadQueueClick(menuItem: MenuItem?) {
        val intent = Intent(this, PhotoReceiverActivity::class.java)
        startActivity(intent)
    }

    fun onWipeCache(menuItem: MenuItem?) {
        _disposables.add(
            Flowable.fromCallable {
                _dataServices.wipeCache()
                true
            }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                    { b: Boolean -> onWipeComplete(b) }
                ) { ex: Throwable? ->
                    stopSyncAnimation()
                    handleApiException(ex)
                }
        )
    }

    fun onReauthenticate(menuItem: MenuItem?) {
        // TODO: check the line below
        //_authStateManager!!.replace(null)

        val intent = Intent(this, LoginActivity::class.java)
        startActivity(intent)
        finish()
    }

    private fun resetNotifications() {
        val mgr = getSystemService(NOTIFICATION_SERVICE) as NotificationManager
        mgr.cancel(0)
        _app!!.notificationCount = 0
    }

    private fun onWipeComplete(b: Boolean) {
        Snackbar.make(binding.toolbar, "Wipe complete.", Snackbar.LENGTH_SHORT).show()
    }

    private fun onItemClicked(
        parent: ExpandableListView,
        v: View,
        groupPosition: Int,
        childPosition: Int,
        id: Long
    ): Boolean {
        val intent: Intent
        val child = _childData[groupPosition][childPosition]
        val type = PhotoListType.valueOf(child[KEY_TYPE]!!)
        if (type == PhotoListType.ByCategory) {
            intent = Intent(parent.context, CategoryListActivity::class.java)
            intent.putExtra("YEAR", child[KEY_NAME]!!.toInt())
        } else {
            intent = Intent(parent.context, PhotoListActivity::class.java)
            intent.putExtra("NAME", child[KEY_NAME])
        }
        intent.putExtra("TYPE", type.toString())
        startActivity(intent)
        return true
    }

    private fun forceSync() {
        startSyncAnimation()
        _disposables.add(
            Flowable.fromCallable { _dataServices.recentCategories }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                    { categories: ApiCollection<Category> -> onSyncComplete(categories) }
                ) { ex: Throwable? ->
                    stopSyncAnimation()
                    handleApiException(ex)
                }
        )
    }

    private fun onSyncComplete(categories: ApiCollection<Category>) {
        stopSyncAnimation()
        val count = categories.count
        if (count == 0L) {
            Snackbar.make(binding.modeExpandableListView, "No updates available.", Snackbar.LENGTH_SHORT)
                .show()
            return
        }

        // TODO: simplify if/when we update api
        /*
        categories
                .stream()
                .filter(x -> !_yearList.contains(x.getYear()))
                .collect(Collectors.toList());
        */
        val newYears: MutableList<Int> = ArrayList()
        for (cat in categories.items) {
            if (!_yearList!!.contains(cat.year)) {
                newYears.add(cat.year)
            }
        }
        var year = "year"
        var cat = "category"
        if (count != 1L) {
            cat = "categories"
        }
        if (newYears.size != 1) {
            year = "years"
        }
        Snackbar.make(
            binding.modeExpandableListView,
            count.toString() + " new " + cat + " found. " + newYears.size + " new " + year + " added.",
            Snackbar.LENGTH_SHORT
        ).show()
        if (newYears.size > 0) {
            prepareYearChildren()
            _adapter!!.notifyDataSetChanged()
        }
    }

    private fun prepareYearChildren() {
        val years = _dataServices.photoYears ?: return

        if (_yearList == null || years.size != _yearList!!.size) {
            _yearList = years
            _yearChildren.clear()
            for (year in _yearList!!) {
                _yearChildren.add(createChild(year.toString(), PhotoListType.ByCategory))
            }
        }
    }

    private fun initModeList() {
        val children: MutableList<Map<String, String?>>

        // by year
        prepareYearChildren()
        _groupData.add(createGroup("Year"))
        _childData.add(_yearChildren)

        // random
        children = ArrayList()
        _groupData.add(createGroup("Random"))
        _childData.add(children)
        children.add(createChild("Surprise Me!", PhotoListType.Random))
        binding.modeExpandableListView.setAdapter(_adapter)
    }

    private fun createGroup(name: String): Map<String, String?> {
        val group: MutableMap<String, String?> = HashMap()
        group[KEY_NAME] = name
        return group
    }

    private fun createChild(name: String, type: PhotoListType): Map<String, String?> {
        val child: MutableMap<String, String?> = HashMap()
        child[KEY_NAME] = name
        child[KEY_TYPE] = type.toString()
        return child
    }

    private fun startSyncAnimation() {
        val iv = layoutInflater.inflate(R.layout.refresh_indicator, binding.toolbar, false) as ImageView
        val rotation = AnimationUtils.loadAnimation(this, R.anim.refresh_rotate)
        rotation.repeatCount = Animation.INFINITE
        iv.startAnimation(rotation)
        _refreshMenuItem!!.actionView = iv
    }

    private fun stopSyncAnimation() {
        val v = _refreshMenuItem!!.actionView
        if (v != null) {
            v.clearAnimation()
            _refreshMenuItem!!.actionView = null
        }
    }

    companion object {
        private const val KEY_NAME = "NAME"
        private const val KEY_TYPE = "TYPE"
        private const val FOUR_HOURS_IN_MILLIS = (4 * 60 * 60 * 1000 // 4 hours
                ).toLong()
    }
}